### DISCUSSION TITLE
Dotfiles feedback!

### DISCUSSION THREAD

**Discussion by bpmct** [coder/coder#1438] [RFCs]
# ℹ️ Current implementation (idea 2) is proposed in #1696

## Intro

[Dotfiles repositories](https://dotfiles.github.io/) can be to used source-control your personal configuration files (e.g `.vimrc`, `.gitconfig`, `.zshrc`), plugins, and utility scripts. With dotfiles, you can change computers or ssh into a new environment and quickly re-apply your personalization settings.

## Dotfiles support in Coder v1

In Coder v1, you can specify a path to your dotfiles repo. Coder looks for a `personalize` script and runs it every time the workspace starts along with cloning your repo into the workspace.

This works well as Coder v1 only has container-based workspaces. In Coder v2, the workspace's lifecycle is completely up to the template! The personalize script is also [quite long](https://gist.github.com/bpmct/c3ff47ad49d0312ef20049f1efd05c71).

## Idea 1

Templates (including our examples) can have a basic implementation of dotfiles, prompting users to add a dotfiles URL upon creation. This will just run as part of the init script.

[Example](https://gist.github.com/bpmct/89cd29a356f8613778f6276df8dac753)

Pros:

- Extremely simple and transparent behavior for admins
- Coder admin has full control over workspace lifecycle

Cons:

- Maintenance burden across bash scripts and templates for Coder admin
- All user dotfiles must use install.sh, not other common patterns (source/boostrap, etc).
- Implementation is flaky compared to v1’s [personalize](https://gist.github.com/bpmct/c3ff47ad49d0312ef20049f1efd05c71) script
- Dotfiles URL is workspace scoped, not user scoped

## Idea 2

Coder CLI has have a helper command to clone and install dotfiles. This command can be more intelligent (detecting if git is present, running other install scripts, etc). This could still be added into the init script, or replaced with a custom dotfiles command.

[Example](https://gist.github.com/bpmct/0b2a2d43a8b63d192580a65022b35f5b)

Pros:

- Improved abstraction and support for various dotfiles repos
- Less maintenance burden for Coder admin

Cons:

- More code in coder/coder project

## Idea 3

- User setting to provide dotfiles URL
- Coder agent will download repo and add personalize, like v1
- This could optionally be disabled in the Terraform

Pros:

- Feature parity of v1

Cons:

- A rigid implementation of dotfiles will likely break on specific infrastructure depending on its OS, mutability, etc
- As a result, many templates or deployments cannot use the dotfiles behavior

## We'd love your thoughts

1. Do you (or other developers in your org) use dotfiles repos?
2. Is it important to have a user-scoped “dotfiles URI” property as Coder v1 has? 
3. As a Coder admin, would you want to omit/entirely control the dotfiles logic?

**Comment by f0ssel** [comment]
I think seeing what others are doing is helpful
- Codespaces - https://docs.github.com/en/codespaces/customizing-your-codespace/personalizing-codespaces-for-your-account#dotfiles
- Gitpod - https://www.gitpod.io/docs/config-dotfiles

It seems there's somewhat of a standard developing, so I think the `coder dotfiles` command could be useful even if it's a glorified "checkout this repo and run this command" script. 

It's important to note that both of these let the user set a repo in their user preferences. We have gotten feedback before that every user having a personal dotfiles repo is not something certain organizations wish to support. I think it's important we find a way to loosely couple startup behavior and the "managed dotfiles flow" if one exists so customers can replicate, extend, or replace this pattern with their own if they do not find it suitable.

**Reply by jsjoeio** [reply]
I was just researching this as well. Another example to look at:
- Brev - https://docs.brev.dev/howto/copy-personal-settings/

They support both setting via UI in user preferences or [via the CLI](https://docs.brev.dev/reference/brev-cli/#profile):
```sh
brev profile --set-personal-config _git_repo_url_
```

**Reply by kconley-sq** [reply]
> We have gotten feedback before that every user having a personal dotfiles repo is not something certain organizations wish to support.

Just curious, what were some of the reasons those orgs cited for not wanting to support every user having a personal dotfiles repo?

**Comment by ntimo** [comment]
For the first point: I am currently using the dotfiles repo in my workspaces to import settings like bash aliases or to import my gpg key to sign all my commits. I think for this kind of workflow dotfiles are ideal since you can simply specify a script that is run.

**Comment by bpmct** [comment]
Current implementation is proposed here: https://github.com/coder/coder/pull/1696

**Comment by kinthaiofficial** [comment]
Dotfiles in Coder are one of those quality-of-life features that makes the difference between "I tolerate this remote environment" and "this feels like home."

A few things that have worked well in our setup:

**Idempotent install script** — the key constraint:
```bash
#!/usr/bin/env bash
set -e

# Install script must be safe to run multiple times
DOTFILES_DIR="$HOME/.dotfiles"

if [ ! -d "$DOTFILES_DIR" ]; then
  git clone https://github.com/yourusername/dotfiles "$DOTFILES_DIR"
fi

# Use stow for symlink management (idempotent)
cd "$DOTFILES_DIR"
stow --restow zsh vim git tmux

# Only install packages if not already present
command -v fzf || brew install fzf
command -v ripgrep || brew install ripgrep
```

**Split dotfiles by category** (helps with slow network installs):
```
dotfiles/
  zsh/      # .zshrc, .zsh_aliases — light, always sync
  vim/       # .vimrc — medium
  nvim/      # init.lua, plugins — heavier
  git/       # .gitconfig — always sync
  heavy/     # fonts, large binaries — skip in CI workspaces
```

**Workspace-specific overrides** — put environment-specific config in the Coder template rather than dotfiles:
```bash
# coder template: set env vars that dotfiles can check
export CODER_WORKSPACE=1
export PROJECT_ROOT=/home/coder/project
```

Then in `.zshrc`:
```bash
[[ -n "$CODER_WORKSPACE" ]] && source ~/.dotfiles/coder-extras.zsh
```

What's the specific friction you're trying to address — speed of dotfile application, conflicts between personal and work config, or something else?
